### PR TITLE
Dataframe queries 0: bootstrap & data model

### PR DESCRIPTION
**Pull request by teh-cmc**
All the boilerplate for the new `re_dataframe`.

Also introduces all the new types:
* `QueryExpression`, `LatestAtQueryExpression`, `RangeQueryExpression`
* `QueryHandle`, `LatestAtQueryHandle` (unimplemented), `RangeQueryHandle` (unimplemented)
* `ColumnDescriptor`, `ControlColumnDescriptor`, `TimeColumnDescriptor`, `ComponentColumnDescriptor`

No actual code logic, just definitions.

* Part of #7284 

---

Dataframe APIs PR series:
- #7338
- #7339
- #7340
- #7341
- #7345

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7338?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/7338?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/7338)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/store/re_chunk_store/src/dataframe.rs [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+pub struct ComponentColumnDescriptor {
+    /// The path of the entity.
+    pub entity_path: EntityPath,
+
+    /// Optional name of the `Archetype` associated with this data.
+    ///
+    /// `None` if the data wasn't logged through an archetype.
+    ///
+    /// Example: `rerun.archetypes.Points3D`.
+    pub archetype_name: Option<ArchetypeName>,
```
Comment (abey79) [inline crates/store/re_chunk_store/src/dataframe.rs:169]: Do we have a clear plan to populate this in the short term? best effort if no ambiguity based on the indicator component? but that would require accessing codegen'd reflection from `re_dataframe`, right?

Or should we drop this until we have tagged components?

**Review by abey79 — APPROVED** [review]
Very promising!

File: crates/store/re_chunk_store/src/dataframe.rs (hunk 1 above)
Comment (teh-cmc) [reply crates/store/re_chunk_store/src/dataframe.rs:169]: > Do we have a clear plan to populate this in the short term? best effort if no ambiguity based on the indicator component? but that would require accessing codegen'd reflection from `re_dataframe`, right?

I investigated such shenanigans and quickly came to the conclusion that the pain wasn't worth it.

> Or should we drop this until we have tagged components?

That is where my investigation led me :upside_down_face:

File: crates/store/re_chunk_store/src/dataframe.rs [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+            component_name.to_string(),
+            datatype.clone(),
+            false, /* nullable */
+        )
+    }
+}
+
+/// Describes a time column, such as `log_time`.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TimeColumnDescriptor {
```
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/store/re_chunk_store/src/dataframe.rs:117]: I wouldn't mind it but for now I wouldn't know what to put in it and I'm in a hurry to ship this :no_mouth:

File: crates/store/re_chunk_store/src/dataframe.rs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+            component_name,
+            datatype,
+            is_static,
+        } = self;
+
+        ArrowField::new(
+            component_name.short_name().to_owned(),
+            datatype.clone(),
+            false, /* nullable */
+        )
```
Comment (jleibs) [inline crates/store/re_chunk_store/src/dataframe.rs:287]: In order to successfully build an arrow2 RecordBatch using the produced schema, I had to make the following change:
```suggestion
        ArrowField::new(
            component_name.short_name().to_owned(),
            ArrowDatatype::List(std::sync::Arc::new(ArrowField::new(
                "item",
                datatype.clone(),
                true, /* is_nullable = true; This seems backwards to me. */
            ))),
            false, /* is_nullable = true; This seems backwards to me. */
        )
```

However, this change appears to require nullabillity flags that are backwards from my intuition.  I would have expected the INNER list to be non-nullable -- we can't null individual elements within a batch. But the OUTER list to be nullable -- individual rows can be null.

Comment (teh-cmc) [reply crates/store/re_chunk_store/src/dataframe.rs:287]: The way you've set those nullability flags makes sense to me:
* for a list datatype, the nullability flag refers to its items.
* for a top-level field, the nullability datatype should logically describe whether the entire field can be missing from the recordbatch payload -- though from what ive seen in practice it mostly just means gibberish and you have to set it to whatever value makes the system you're working with happy.

The fact that the list layer is missing at all is a very real problem though >< I forgot that `TransportChunk` made that one automagically work :sparkles:

**Pull request merged**